torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+from contextlib import contextmanager
+from dataclasses import dataclass
+from enum import Enum
+
+from typing import Any, cast, Dict, List, Optional, Tuple
+
+import torch
+import torch.fx as fx
+import torch.library
+import torch.nn as nn
+
+import torch.utils._pytree as pytree
+
+from torch.distributed._tensor import DeviceMesh, distribute_tensor, Replicate, Shard
+
+from torch.distributed._tensor._utils import compute_local_shape
+from torch.distributed._tensor.op_schema import PlacementStrategy, StrategyList
+from torch.distributed._tensor.placement_types import _Partial, DTensorSpec, Placement
+from torch.distributed._tensor.redistribute import _redistribute_with_local_tensor
+from torch.fx import GraphModule
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch.fx.passes.shape_prop import _extract_tensor_metadata
+from torch.nn.utils._named_member_accessor import NamedMemberAccessor
+
+
+aten = torch.ops.aten
+
+# Dummy op used by data parallel to tag gradients.
+_spmd_lib_def = torch.library.Library("_spmd", "DEF")
+_spmd_lib_def.define("tag_grad(Tensor self) -> Tensor")
+
+_spmd_lib_impl = torch.library.Library("_spmd", "IMPL")
+for dispatch_key in ("CPU", "CUDA", "Meta"):
+    _spmd_lib_impl.impl("tag_grad", lambda x: x, dispatch_key)
```

Comment:
Instead of registering each dispatch key separately, you can register this to CompositeExplicitAutograd